File: Problem Set 6/credit.py
# Made by siraponte

from cs50 import get_string

# Take card number as string to highligt digits easly
n = get_string("Credit Card Number: ")

# Check if credit card number's length is right
if (int(n) != 13 or int(n) != 15 or int(n) != 16):
    
    print("INVALID")

# Sum_ to take sum of digits
sum_ = 0

# To jump between odd and even digits
# Even and odd refer to position , starting from the right 
jump_ = 0

# Loop through digits
for i in range(len(n) + 1):
    
    # Takes and multiply even digits
    double = int(n[-i]) * 2
    
    # Takes odd digits
    not_double = int(n[-i])
    
    # Check if i is equal 1 (With 0 can't take last digits cause -0 does not exist)
    if i > 0:
        
        # If jump is equal 0 then the digit is odd
        if jump_ == 0:
            
            # Adds the digit to sum
            sum_ += not_double
            
            # After an odd digit , comes an even digit
            jump_ = 1
            
        # Else is even
        else:
            
            # Check if the double of the digit has 1 or 2 digits
            if double > 9:
                
                # If it has 2 digits , take double as string to separate them easly
                s_double = str(double)
                
                # And adds to sum the separate digits es.(12 -> 1 + 2)
                sum_ += int(s_double[0]) + int(s_double[1])
                
                # After an even digit , comes an odd digit
                jump_ = 0
                
            else:
                
                # If it has 1 digit , adds to sum
                sum_ += double
                
                # After an even digit , comes an odd digit
                jump_ = 0

# Check if final digit of sum is 0
if ((sum_ % 10) == 0):
    
    # Check for credit card number's length
    if len(n) == 13:
        
        # Check if card's number start with 4 
        if (int(n[0]) == 4):
            
            print("VISA")
            
        else:
            
            print("INVALID")
    
    elif len(n) == 15:
        
        # Check if card's number start with 34 or 37
        if (int(n[0:2]) == 34 or int(n[0:2]) == 37):
            
            print("AMEX")
            
        else:
            
            print("INVALID")
    
    elif len(n) == 16:
        
        # Check if card's number start with 4
        if (int(n[0]) == 4):
            
            print("VISA")
        
        # Check if card's number start with 51,52,53,54 or 55
        elif (int(n[0:2]) == 51 or int(n[0:2]) == 52 or int(n[0:2]) == 53 or int(n[0:2]) == 54 or int(n[0:2]) == 55):
            
            print("MASTERCARD")
        
        else:
            
            print("INVALID")

File: Problem Set 1/credit.c
//Made by siraponte

//Libraries
#include <stdio.h>
#include <cs50.h>
#include <math.h>

//Main function
int main(void)
{
    
    int long n = get_long("Credit Card Number : ");
    int long c, c2, d1 = 1, d10 = 10;
    
    int l = 1 + log10(n);
    int s = 0;
    
    //Function to take 16 digits cards
    if (l % 2 == 0)
    {
        //Loop to sum odd digits
        for (int i = 0 ; i < l / 2 ; i++)
        {
            c = ((n / d1) % 10);
            
            s += c;
            
            d1 *= 100;
        }
        //Loop to sum even digits
        for (int i2 = 0 ; i2 < l / 2 ; i2++)
        {
            c2 = (((n / d10) % 10) * 2);
            
            //Function to separate double digits numbers
            if (c2 > 9)
            {
                s += ((c2 % 10) + ((c2 / 10) % 10));
            }
            else
            {
                s += c2;
            }
            d10 *= 100;
        }
    }
    //Function to take 13 and 15 digits cards
    else
    {
        //Loop to sum odd digits
        for (int i3 = 0 ; i3 < l / 2 + 1 ; i3++)
        {
            c = ((n / d1) % 10);
            
            s += c;
            
            d1 *= 100;
        }
        //Loop to sum even digits numbers
        for (int i4 = 0 ; i4 < l / 2 ; i4++)
        {
            c2 = (((n / d10) % 10) * 2);
            
            //Function to separate double digits
            if (c2 > 9)
            {
                s += ((c2 % 10) + ((c2 / 10) % 10));
            }
            else
            {
                s += c2;
            }
            d10 *= 100;
        }
    }
    
    int long d12 = 1000000000000, d13 = 10000000000000, d14 = 100000000000000, d15 = 1000000000000000;
    
    int g12 = (n / d12), g13 = (n / d13), g14 = (n / d14), g15 = (n / d15);
    
    //Function to print the right cards' names
    if ((s % 10) == 0)
    {
        if (l == 13)
        {
            if (g12 == 4)
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else if (l == 15)
        {
            if ((g13 == 34) || (g13 == 37))
            {
                printf("AMEX\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else if (l == 16)
        {
            if (g15 == 4)
            {
                printf("VISA\n");
            }
            else if ((g14 == 51) || (g14 == 52) || (g14 == 53) || (g14 == 54) || (g14 == 55))
            {
                printf("MASTERCARD\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
            
        }
    }
    else
    {
        printf("INVALID\n");
    }
}
